File: stock_logic_poc.py
# Logic for Stock Automation - Savia App
# This script simulates the backend logic for processing a production batch

class StockManager:
    def __init__(self, db_state):
        self.db = db_state

    def simulate_production(self, producto_id, cantidad):
        print(f"\n--- Iniciando Producción: {cantidad} unidades de Producto ID {producto_id} ---")
        
        # 1. Get Formula
        ingredients_needed = self.db['formulas'].get(producto_id, [])
        if not ingredients_needed:
            return "Error: No se encontró fórmula para este producto."

        # 2. Check Stock Availability
        required_amounts = []
        for insumo_id, qty_per_unit in ingredients_needed:
            total_required = qty_per_unit * cantidad
            current_stock = self.db['insumos'][insumo_id]['stock']
            
            if current_stock < total_required:
                insumo_name = self.db['insumos'][insumo_id]['nombre']
                return f"Error: Stock insuficiente de '{insumo_name}'. Necesario: {total_required}, Disponible: {current_stock}"
            
            required_amounts.append((insumo_id, total_required))

        # 3. Process Transaction (Atomic simulation)
        for insumo_id, amount in required_amounts:
            self.db['insumos'][insumo_id]['stock'] -= amount
            print(f"Descontado {amount} de Insumo ID {insumo_id}")

        self.db['productos'][producto_id]['stock'] += cantidad
        print(f"Sumado {cantidad} a Produto ID {producto_id}")
        
        return "Éxito: Producción completada y stock actualizado."

# Mock Database State
mock_db = {
    'insumos': {
        1: {'nombre': 'Aceite de Coco', 'stock': 1000},
        2: {'nombre': 'Cera de Abeja', 'stock': 500}
    },
    'productos': {
        1: {'nombre': 'Bálsamo Labial', 'stock': 0}
    },
    'formulas': {
        1: [(1, 10), (2, 5)] # (InsumoID, Qty)
    }
}

manager = StockManager(mock_db)

# Test 1: Successful Production
print(manager.simulate_production(1, 50))

# Test 2: Insufficient Stock
print(manager.simulate_production(1, 200))

File: stock_logic_poc.js
/**
 * Logic for Stock Automation - Savia App
 * This script simulates the backend logic for processing a production batch in Node.js
 */

class StockManager {
    constructor(dbState) {
        this.db = dbState;
    }

    simulateProduction(productoId, cantidad) {
        console.log(`\n--- Iniciando Producción: ${cantidad} unidades de Producto ID ${productoId} ---`);
        
        // 1. Get Formula
        const ingredientsNeeded = this.db.formulas[productoId];
        if (!ingredientsNeeded) {
            return "Error: No se encontró fórmula para este producto.";
        }

        // 2. Check Stock Availability
        const requiredAmounts = [];
        for (const { insumoId, qtyPerUnit } of ingredientsNeeded) {
            const totalRequired = qtyPerUnit * cantidad;
            const currentStock = this.db.insumos[insumoId].stock;
            
            if (currentStock < totalRequired) {
                const insumoName = this.db.insumos[insumoId].nombre;
                return `Error: Stock insuficiente de '${insumoName}'. Necesario: ${totalRequired}, Disponible: ${currentStock}`;
            }
            
            requiredAmounts.push({ insumoId, totalRequired });
        }

        // 3. Process Transaction (Atomic simulation)
        requiredAmounts.forEach(({ insumoId, totalRequired }) => {
            this.db.insumos[insumoId].stock -= totalRequired;
            console.log(`Descontado ${totalRequired} de Insumo ID ${insumoId}`);
        });

        this.db.productos[productoId].stock += cantidad;
        console.log(`Sumado ${cantidad} a Producto ID ${productoId}`);
        
        return "Éxito: Producción completada y stock actualizado.";
    }
}

// Mock Database State
const mockDb = {
    insumos: {
        1: { nombre: 'Aceite de Coco', stock: 1000 },
        2: { nombre: 'Cera de Abeja', stock: 500 }
    },
    productos: {
        1: { nombre: 'Bálsamo Labial', stock: 0 }
    },
    formulas: {
        1: [
            { insumoId: 1, qtyPerUnit: 10 }, 
            { insumoId: 2, qtyPerUnit: 5 }
        ]
    }
};

const manager = new StockManager(mockDb);

// Test 1: Successful Production
console.log(manager.simulateProduction(1, 50));

// Test 2: Insufficient Stock
console.log(manager.simulateProduction(1, 200));
